client/utils/src/notification.rs

```diff
@@ -0,0 +1,174 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021-2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Provides mpsc notification channel that can be instantiated
+//! _after_ it's been shared to the consumer and producers entities.
+//!
+//! Useful when building RPC extensions where, at service definition time, we
+//! don't know whether the specific interface where the RPC extension will be
+//! exposed is safe or not and we want to lazily build the RPC extension
+//! whenever we bind the service to an interface.
+//!
+//! See [`sc-service::builder::RpcExtensionBuilder`] for more details.
+
+use std::{marker::PhantomData, sync::Arc};
+
+use crate::mpsc::{tracing_unbounded, TracingUnboundedReceiver, TracingUnboundedSender};
+
+use parking_lot::Mutex;
+
+/// Collection of channel sending endpoints shared with the receiver side
+/// so they can register themselves.
+type SharedSenders<Payload> = Arc<Mutex<Vec<TracingUnboundedSender<Payload>>>>;
+
+/// Trait used to define the "tracing key" string used to tag
+/// and identify the mpsc channels;
+pub trait TracingKeyStr {
+	/// Const `str` representing the "tracing key" used to tag and identify
+	/// the mpsc channels owned by the object implemeting this trait;
+	const TRACING_KEY: &'static str;
+}
+
+/// The sending half of the notifications channel(s).
+///
+/// Used to send notifications from the BEEFY gadget side.
+pub struct NotificationSender<Payload: Clone, Error> {
+	subscribers: SharedSenders<Payload>,
+	_err: PhantomData<Error>,
+}
+
+// Because of https://github.com/rust-lang/rust/issues/26925 we need to
+// manually implement `clone` here.
+impl<Payload: Clone, Error> Clone for NotificationSender<Payload, Error> {
+	fn clone(&self) -> Self {
+		Self { subscribers: self.subscribers.clone(), _err: PhantomData }
+	}
+}
+
+impl<Payload: Clone, Error> NotificationSender<Payload, Error> {
+	/// The `subscribers` should be shared with a corresponding `NotificationStream`.
+	fn new(subscribers: SharedSenders<Payload>) -> Self {
+		Self { subscribers, _err: PhantomData }
+	}
+
+	/// Send out a notification to all subscribers that a new payload is available for a
+	/// block.
+	pub fn notify(&self, payload: impl FnOnce() -> Result<Payload, Error>) -> Result<(), Error> {
```

Comment:
Would it make sense to only have the generic `Error` type on this function rather than on the struct itself? Given that we only deal with errors when creating the payload.

Reply:
good point. moved the generic to the function and simplified struct definitions